### ISSUE TITLE
Function called from macro_export macro is warned as as dead code

### ISSUE THREAD

**Issue by btzy** [rust-lang/rust#118287]
Consider the following code:

```rust
fn my_func() -> i32 {
    1
}

#[macro_export]
macro_rules! my_macro {
    () => {
        my_func();
    };
}
```

Since `my_macro()` is exported, it is callable from other modules, and so it is not dead code.  It then follows that `my_func()`, which is called by the macro, is also not dead code.  However, we get a dead code warning for `my_func()`:
```
warning: function `my_func` is never used
 --> src/lib.rs:1:4
  |
1 | fn my_func() -> i32 {
  |    ^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

Link to playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1bb729e143d43fab183175a6fefd2de3

Occurs in both stable (1.74.0) and nightly (1.76.0-nightly (2023-11-24 37b2813a7be580ef5904)) on playground.

**Comment by SkiFire13** [comment]
> It then follows that `my_func()`, which is called by the macro, is also not dead code.

This is not true, `my_func` also needs to be visible from the outside in order to be called through `my_macro`. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f462817fb2c751febf09317f7132fed1

**Comment by btzy** [comment]
My understanding of Rust was incorrect.
